hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -116,33 +158,52 @@ public List<R> consume(
 		final PersistenceContext persistenceContext = session.getPersistenceContext();
 		final TypeConfiguration typeConfiguration = session.getTypeConfiguration();
 		final QueryOptions queryOptions = rowProcessingState.getQueryOptions();
-
 		RuntimeException ex = null;
 		try {
 			persistenceContext.getLoadContexts().register( jdbcValuesSourceProcessingState );
 
-			final List<R> results = new ArrayList<>();
-
 			final JavaType<R> domainResultJavaType = resolveDomainResultJavaType(
 					rowReader.getDomainResultResultJavaType(),
 					rowReader.getResultJavaTypes(),
 					typeConfiguration
 			);
 
-			final ResultHandler<R> resultHandlerToUse;
+			final boolean isEnityResultType = domainResultJavaType instanceof EntityJavaType;
 
-			if ( uniqueSemantic == UniqueSemantic.ALLOW
-					&& domainResultJavaType instanceof EntityJavaType ) {
-				resultHandlerToUse = ListResultsConsumer::deDuplicationHandling;
+			final Results<R> results;
+			if ( ( uniqueSemantic == UniqueSemantic.ALLOW || uniqueSemantic == UniqueSemantic.FILTER ) && isEnityResultType ) {
+				results = new EntityResult<>( domainResultJavaType );
 			}
 			else {
-				resultHandlerToUse = this.resultHandler;
+				results = new Results<>( domainResultJavaType );
 			}
 
-			while ( rowProcessingState.next() ) {
-				final R row = rowReader.readRow( rowProcessingState, processingOptions );
-				resultHandlerToUse.handle( row, domainResultJavaType, results, rowProcessingState );
-				rowProcessingState.finishRowProcessing();
+			if ( this.uniqueSemantic == UniqueSemantic.FILTER || this.uniqueSemantic == UniqueSemantic.ALLOW && isEnityResultType ) {
+				while ( rowProcessingState.next() ) {
+					results.addUnique( rowReader.readRow( rowProcessingState, processingOptions ) );
+					rowProcessingState.finishRowProcessing();
+				}
+			}
+			else if ( this.uniqueSemantic == UniqueSemantic.ASSERT ) {
+				while ( rowProcessingState.next() ) {
+					if ( !results.addUnique( rowReader.readRow( rowProcessingState, processingOptions ) )
+							&& !rowProcessingState.hasCollectionInitializers ) {
```

Comment:
When `uniqueSemantics == ASSERT && rowProcessingState.hasCollectionInitializers` the effect is the same as `FILTER` in the above if-block. IMO we should add that condition to the upper if, like:

```
if ( uniqueSemantic == UniqueSemantic.FILTER || uniqueSemantics == ASSERT && rowProcessingState.hasCollectionInitializers || uniqueSemantic == UniqueSemantic.ALLOW && isEnityResultType ) {
```

and remove the check for `&& !rowProcessingState.hasCollectionInitializers` here.